mlsdbtool/main.c

```diff
@@ -0,0 +1,180 @@
+        network += 3;
+        break;
+    }
+}
+
+int main() {
+    char c;
+    char line[150];
+    char mcc_fn[16];
+    char loc_fn[16];
+    size_t pos = 0, count = 0, mcc_old = 0, mcc_num = 0, mcc_p = 0, net_p = 0, area_p = 0, cell_p = 0, lon_p = 0, lat_p = 0;
+    FILE *fp_mcc = NULL, *fp_loc = NULL;
+    while (read(STDIN, &c, 1) > 0) {
+        if (c == NEWLINE) {
+            pos = 0;
+            count = 0;
+            position = 0;
+            network = 0;
+            mcc_num = atoi(&line[mcc_p]);
+            // We have a new mcc; we need to open new files.
+            if (mcc_num != mcc_old) {
+                if (mcc_old > 0) {
+                    fclose(fp_mcc);
+                    fclose(fp_loc);
+                }
+                strcpy(mcc_fn, "./");
+                strcpy(loc_fn, "./");
+                strcat(mcc_fn, &line[mcc_p]);
+                strcat(loc_fn, &line[mcc_p]);
+                strcat(mcc_fn, ".ntw");
+                strcat(loc_fn, ".loc");
+                fp_mcc = fopen(mcc_fn, "w");
+                fp_loc = fopen(loc_fn, "w");
+                if (fp_mcc == NULL || fp_loc == NULL) {
+                    fprintf(stderr, "Unable to open outfile for mcc %ld.\n", mcc_num);
+                    return 1;
+                }
+                mcc_old = mcc_num;
+                previous = 0;
+            }
+            add_network(&line[net_p], 46);
+            add_network(&line[area_p], 30);
+            add_network(&line[cell_p], 2);
+            add_radio(line);
+            if (previous > network) {
+                fprintf(stderr, "ERROR: The current record has value %lu, which is smaller than the previous %lu\n", network, previous);
+                fprintf(stderr, "The reader will not be able to properly search a file that isn't sorted correctly. Please check your sort.\n");
+                return 1;
+            }
+            previous = network;
+            add_position(&line[lon_p], 32);
+            add_position(&line[lat_p], 0);
+            fwrite(&network, sizeof(network), 1, fp_mcc);
+            fwrite(&position, sizeof(position), 1, fp_loc);
+        }
+        else if (c == ',') {
+            line[pos] = '\0'; // Null instead of comma for later string operations.
+            ++count;
+            ++pos;
+            switch(count) {
+            case 1:
+                mcc_p = pos;
+                break;
+            case 2:
+                net_p = pos;
+                break;
+            case 3:
+                area_p = pos;
+                break;
+            case 4:
+                cell_p = pos;
+                break;
+            case 6:
+                lon_p = pos;
+                break;
+            case 7:
+                lat_p = pos;
+                break;
+            default:
+                break;
+            }
+        }
+        else {
+            line[pos] = c;
+            ++pos;
+        }
+    }
+    fclose(fp_mcc);
+    fclose(fp_loc);
```

Comment:
I think `fclose()` is not NULL-safe but anyways app is terminating after this.

Reply:
Also NULL check is handled immediately after fopen is called (line 126)

Reply:
Line 126 may not be reached. Execute tool, ctrl+d -> crash.